sdks/java/core/src/main/java/org/apache/beam/sdk/util/WatermarkHold.java

```diff
@@ -264,87 +260,140 @@ private Instant addElementHold(ReduceFn<?, ?, ?, W>.ProcessValueContext context)
+      // If paneIsEmpty then this hold is just for empty ON_TIME panes, so we want to keep
+      // the hold away from the combining function in elementHoldTag.
+      // However if !paneIsEmpty then it could make sense  to use the elementHoldTag here.
+      // Alas, onMerge is forced to add an end of window or garbage collection hold without
+      // knowing whether an element hold is already in place (stopping to check is too expensive).
+      // This it would end up adding an element hold at the end of the window which could
+      // upset the elementHoldTag combining function.
       context.state().access(EXTRA_HOLD_TAG).add(eowHold);
     }
     WindowTracing.trace(
         "WatermarkHold.addEndOfWindowHold: end-of-window hold at {} is {} for "
         + "key:{}; window:{}; inputWatermark:{}; outputWatermark:{}",
-        eowHold, which, context.key(), context.window(), inputWM,
-        outputWM);
+        eowHold, which, context.key(), context.window(), inputWM, outputWM);
 
-    return tooLate ? null : eowHold;
+    return added ? eowHold : null;
   }
 
   /**
-   * Add a garbage collection hold, if required. Return the {@link Instant} at which hold was added,
-   * or {@literal null} if no hold was added.
-   *
-   * <p>The garbage collection hold gurantees that any empty final pane can be given
-   * a timestamp which will not be considered beyond allowed lateness by any downstream
-   * computation. If we are sure no empty final panes can be emitted then there's no need
-   * for an additional hold.
+   * Attempt to add a 'garbage collection hold' if it is required. Return the {@link Instant} at
+   * which the hold was added (ie the end of window time plus allowed lateness), or
+   * {@literal null} if no hold was added. We only add the hold if it is distinct from
+   * what would be added by {@link #addEndOfWindowHold}. In other words,
+   * {@link WindowingStrategy#getAllowedLateness} must be non-zero.
+   * <p>A garbage collection hold is added in two situations:
+   * <ol>
+   * <li>An incoming element came in behind the output watermark, and was too late for placing
+   * the usual element hold or an end of window hold. Place the garbage collection hold so that
+   * we can guarantee when the pane is finally triggered its output will not be dropped due to
+   * excessive lateness by any downstream computation.
+   * <li>The {@link WindowingStrategy#getClosingBehavior()} is {@link ClosingBehavior#FIRE_ALWAYS},
+   * and thus we guarantee a final pane will be emitted for all windows (except
+   * the {@link GlobalWindow}, see below) which saw at least one element. Again, the garbage
+   * collection hold guarantees that any empty final pane can be given a timestamp which will
+   * not be considered beyond allowed lateness by any downstream computation.
+   * </ol>
+   * <p>If we are in the {@link GlobalWindow} then case 1 still applies so that panes will
```

Comment:
This implied behavior has gone away with the getOnTimeBehavior, so the user must explicitly opt-out of the empty final pane. So I think the comment is now straightforward.